### PR TITLE
issue-46477 Fix issue about checking last char in datetime string.

### LINKED ISSUE #46477: DateTime.Parse ISO 8601 allowing extra arbitrary character at the end of the date
When parsing an ISO 8601 DateTime the parser allows 1 single extra arbitrary character 

The following dates are all parsed correctly 

2020-5-7T09:37:00.0000000-07:00

2020-5-7T09:37:00.0000000-07:00x

2020-5-7T09:37:00.0000000-07:00@

In summary 

1. When parsing in `ParseISO8601` it grabs the correct *timezone* information `-07:00` in `ParseTimeZone`
2. This leaves the `str` with 1 last index
3. It then checks for `str.Match('#')` which *prefix increments* the index (which I believe is the fault). 
4. At this point it assumes the parsing is complete in subsequent checks due to the index being at the end of string

This seems to be a problem in 

````
//
// Parse the ISO8601 format string found during Parse();
//
//
private static bool ParseISO8601(ref DateTimeRawInfo raw, ref __DTString str, DateTimeStyles styles, ref DateTimeResult result) {
````

In particular, when given a valid *ISO 8601* `DateTime`,  `ParseTimeZone` will return the correct timezone, and at this point  `str` has 1 valid index left (the arbitrary character)

It then checks for a _hash_ with `str.Match('#')` 

````
   ...
   str.SkipWhiteSpaces();
   if (str.Match('#'))
   {
      if (!VerifyValidPunctuation(ref str))
      {
         result.SetFailure(ParseFailureKind.Format, "Format_BadDateTime", null);
         return false;
      }
      str.SkipWhiteSpaces();
   }
   if (str.Match('\0'))
   {
      if (!VerifyValidPunctuation(ref str))
      {
         result.SetFailure(ParseFailureKind.Format, "Format_BadDateTime", null);
         return false;
      }
   }
   if (str.GetNext())
   {
      // If this is true, there were non-white space characters remaining in the DateTime
      result.SetFailure(ParseFailureKind.Format, "Format_BadDateTime", null);
      return false;
   }
}
````

Since `Match` _prefix increments_ the index and compares it with the valid length, the check fails, however the string is still incremented. 

````
internal bool Match(char ch)
{
   if (++Index >= len)
   {
      return (false);
   }
   if (Value[Index] == ch)
   {
      m_current = ch;
      return (true);
   }
   Index--;
   return (false);
}
````

The result of this allows the date to pass the next 2 checks on  `str.Match('\0')` and `str.GetNext()` without causing an error. 

What it seems like its designed to do, is check the _**current**_ character for _hash_, then _null termination_, then fail if there is anything else left.

Note : this is not a problem with 2 arbitrary characters, as the index is decremented back to the original and the checks perform as expected.

### PR DESCRIPTION
**Pull request by lateapexearlyspeed**
Fix #46477

### CODE REVIEW

File: src/libraries/System.Runtime/tests/System/DateTimeTests.cs [hunk 1]
```diff
@@ -1041,6 +1041,10 @@ public static void Parse_InvalidArguments_Throws()
             Assert.Throws<FormatException>(() => DateTime.Parse(""));
             Assert.Throws<FormatException>(() => DateTime.Parse("", new MyFormatter()));
             Assert.Throws<FormatException>(() => DateTime.Parse("", new MyFormatter(), DateTimeStyles.NoCurrentDateDefault));
+
+            Assert.Throws<FormatException>(() => DateTime.Parse("2020-5-7T09:37:00.0000000-07:00c"));
+            Assert.Throws<FormatException>(() => DateTime.Parse("2020-5-7T09:37:00.0000000-07:00c", new MyFormatter()));
+            Assert.Throws<FormatException>(() => DateTime.Parse("2020-5-7T09:37:00.0000000-07:00c", new MyFormatter(), DateTimeStyles.NoCurrentDateDefault));
```
Comment (danmosemsft) [inline src/libraries/System.Runtime/tests/System/DateTimeTests.cs:1047]: these should continue to work (per the strange treatment of # and \0), so you might want to add a test to prove it was not broken by your change:
`2020-5-7T09:37:00.0000000+00:00\0`
`#2020-5-7T09:37:00.0000000+00:00#`
`#2020-5-7T09:37:00.0000000+00:00#\0`
I only see tests for these without the timezone part.
These should still fail
`2020-5-7T09:37:00.0000000+00:00#`
`2020-5-7T09:37:00.0000000+00:00#\0`

Comment (tarekgh) [reply src/libraries/System.Runtime/tests/System/DateTimeTests.cs:1047]: I second @danmosemsft for adding more test cases for valid formats. Also, please ensure adding cases when there is nothing at the end too. something like `2020-5-7T09:37:00.0000000-07:00`

**Comment by tarekgh** [comment]
@lateapexearlyspeed did you have a chance to address the feedback on this PR?

**Comment by lateapexearlyspeed** [comment]
@tarekgh Sorry I am little busy these days, will fix those comment these days.

**Comment by tarekgh** [comment]
@lateapexearlyspeed no problem and thanks again for helping with this one.

File: src/libraries/System.Runtime/tests/System/DateTimeTests.cs (hunk 1 above)
Comment (lateapexearlyspeed) [reply src/libraries/System.Runtime/tests/System/DateTimeTests.cs:1047]: @danmosemsft @tarekgh Fixed, please review, thanks.

**Comment by tarekgh** [comment]
The failure in this PR is unrelated and tracked by the linked issue #32805

**Review by tarekgh — APPROVED** [review]
Thanks @lateapexearlyspeed for getting this fixed.

**Pull request merged**